File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/BuiltinFunctions/Add.cpp
#include "Add.h"
#include "../Code/ExpressionType.h"
#include "../Code/FunctionUtils.h"

#include <limits>
#include <any>
#include <cstdlib>

AdaptiveExpressions_BuiltinFunctions::Add::Add() : 
    ExpressionEvaluatorWithArgs(ExpressionType::Add, (ReturnType)((int)ReturnType::String | (int)ReturnType::Number))
{
}

ValueErrorTuple AdaptiveExpressions_BuiltinFunctions::Add::EvaluateOperator(std::vector<std::any> args)
{
    std::any result;
    std::string error;
    auto firstItem = args[0];
    auto secondItem = args[1];

    bool stringConcat = !FunctionUtils::isNumber(firstItem) || !FunctionUtils::isNumber(secondItem);

    if ((!firstItem.has_value() && FunctionUtils::isNumber(secondItem))
        || (!secondItem.has_value() && FunctionUtils::isNumber(firstItem)))
    {
        error = "Operator '+' or add cannot be applied to operands of type 'number' and null object.";
    }
    else
    {
        if (stringConcat)
        {
            result = "{firstItem?.ToString()}{secondItem?.ToString()}";
        }
        else
        {
            result = EvalAdd(args[0], args[1]);
        }
    }

    return ValueErrorTuple(result, error);
}

std::any AdaptiveExpressions_BuiltinFunctions::Add::EvalAdd(std::any a, std::any b)
{
    if (a.has_value())
    {
        // throw new ArgumentNullException(nameof(a));
    }

    if (b.has_value())
    {
        // throw new ArgumentNullException(nameof(b));
    }

    bool valueASuccess{}, valueBSuccess{};
    if (FunctionUtils::isInteger(a) && FunctionUtils::isInteger(b))
    {
        int valueA = FunctionUtils::castToType<int>(a, valueASuccess);
        int valueB = FunctionUtils::castToType<int>(b, valueBSuccess);
        return valueA + valueB;
    }

    double valueA = FunctionUtils::castToType<double>(a, valueASuccess);
    double valueB = FunctionUtils::castToType<double>(b, valueBSuccess);
    return (valueA + valueB);
}

void AdaptiveExpressions_BuiltinFunctions::Add::ValidateExpression(Expression* expression)
{
    FunctionUtils::ValidateArityAndAnyType(expression, 2, INT_MAX, ReturnType::String | ReturnType::Number);
}

File: libraries/AdaptiveExpressionsSTJ/BuiltinFunctions/Add.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Globalization;

namespace AdaptiveExpressions.BuiltinFunctions
{
    /// <summary>
    /// Return the result from adding two or more numbers (pure number case) or concatting two or more strings (other case).
    /// </summary>
    internal class Add : ExpressionEvaluator
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="Add"/> class.
        /// </summary>
        public Add()
            : base(ExpressionType.Add, Evaluator(), ReturnType.String | ReturnType.Number, Validator)
        {
        }

        private static EvaluateExpressionDelegate Evaluator()
        {
            return FunctionUtils.ApplySequenceWithError(
                    args =>
                    {
                        object result = null;
                        string error = null;
                        var firstItem = args[0];
                        var secondItem = args[1];
                        var stringConcat = !firstItem.IsNumber() || !secondItem.IsNumber();

                        if ((firstItem == null && secondItem.IsNumber())
                            || (secondItem == null && firstItem.IsNumber()))
                        {
                            error = "Operator '+' or add cannot be applied to operands of type 'number' and null object.";
                        }
                        else
                        {
                            if (stringConcat)
                            {
                                result = $"{firstItem?.ToString()}{secondItem?.ToString()}";
                            }
                            else
                            {
                                result = EvalAdd(args[0], args[1]);
                            }
                        }

                        return (result, error);
                    }, FunctionUtils.VerifyNumberOrStringOrNull);
        }

        private static object EvalAdd(object a, object b)
        {
            if (a == null)
            {
                throw new ArgumentNullException(nameof(a));
            }

            if (b == null)
            {
                throw new ArgumentNullException(nameof(b));
            }

            if (a.IsInteger() && b.IsInteger())
            {
                return Convert.ToInt64(a, CultureInfo.InvariantCulture) + Convert.ToInt64(b, CultureInfo.InvariantCulture);
            }

            return FunctionUtils.CultureInvariantDoubleConvert(a) + FunctionUtils.CultureInvariantDoubleConvert(b);
        }

        private static void Validator(Expression expression)
        {
            FunctionUtils.ValidateArityAndAnyType(expression, 2, int.MaxValue, ReturnType.String | ReturnType.Number);
        }
    }
}
